pkg/verification/verification_service.go

```diff
@@ -147,56 +189,78 @@ func generateVerificationCode() (string, error) {
 		// Set a new timestamp
 		signup.Annotations[v1alpha1.UserSignupVerificationTimestampAnnotationKey] = now.Format(TimestampLayout)
 		signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = "0"
 	}
 
-	attemptsMade, err := strconv.Atoi(signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey])
-	if err != nil {
-		// We shouldn't get an error here, but if we do, we should probably set verification attempts to max allowed
-		// so that we at least now have a valid value
-		signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(s.config.GetVerificationAttemptsAllowed())
-		return signup, err
+	attemptsMade, convErr := strconv.Atoi(signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey])
+	if convErr != nil {
+		// We shouldn't get an error here, but if we do, we will set verification attempts to max allowed
+		// so that we at least now have a valid value, and let the workflow continue to the
+		// subsequent attempts check
+		attemptsMade = s.config.GetVerificationAttemptsAllowed()
+		signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(attemptsMade)
 	}
 
 	// If the user has made more attempts than is allowed per day, return an error
 	if attemptsMade >= s.config.GetVerificationAttemptsAllowed() {
-		return signup, errors.NewTooManyRequestsError("too many verification attempts", "")
+		verificationErr = errors.NewTooManyRequestsError("too many verification attempts", "")
 	}
 
-	exp, err := time.Parse(TimestampLayout, signup.Annotations[v1alpha1.UserVerificationExpiryAnnotationKey])
-	if err != nil {
-		// If the verification expiry timestamp is corrupt or missing, then return an error
-		return signup, errors.NewInternalError(err, "error parsing expiry timestamp")
+	if verificationErr == nil {
+		// Parse the verification expiry timestamp
+		exp, parseErr := time.Parse(TimestampLayout, signup.Annotations[v1alpha1.UserVerificationExpiryAnnotationKey])
+		if parseErr != nil {
+			// If the verification expiry timestamp is corrupt or missing, then return an error
+			verificationErr = errors.NewInternalError(parseErr, "error parsing expiry timestamp")
+		} else if now.After(exp) {
+			// If it is now past the expiry timestamp for the verification code, return a 403 Forbidden error
+			verificationErr = errors.NewForbiddenError("expired", "verification code expired")
+		}
 	}
 
-	if now.After(exp) {
-		// If it is now past the expiry timestamp for the verification code, return a 403 Forbidden error
-		return signup, errors.NewForbiddenError("expired", "verification code expired")
+	if verificationErr == nil {
+		if code != signup.Annotations[v1alpha1.UserSignupVerificationCodeAnnotationKey] {
+			// The code doesn't match
+			attemptsMade++
+			signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(attemptsMade)
+			verificationErr = errors.NewForbiddenError("invalid code", "the provided code is invalid")
```

Comment:
without returning the error and overwriting the error, we not only loose the previous error but we also continue on as though no error happened - is this intended?